cluster-autoscaler/cloudprovider/gce/cache.go

```diff
@@ -0,0 +1,258 @@
+	"strings"
+	"sync"
+	"time"
+
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+
+	"github.com/golang/glog"
+	gce "google.golang.org/api/compute/v1"
+)
+
+// GceCache is used for caching cluster resources state.
+type GceCache struct {
+	// Cache content.
+	migs            []*MigInformation
+	migCache        map[GceRef]Mig
+	resourceLimiter *cloudprovider.ResourceLimiter
+	machinesCache   map[MachineTypeKey]*gce.MachineType
+	// Freshness timestamps.
+	machinesCacheLastRefresh time.Time
+	// Locks. Rules of locking:
+	// - migsMutex protects only migs.
+	// - cacheMutex protects migCache, resourceLimiter and machinesCache.
+	// - methods that take cacheMutex may call methods that take migsMutex.
+	// - methods that take migsMutex may not call methods that take cacheMutex.
+	cacheMutex sync.Mutex
+	migsMutex  sync.Mutex
+	// Service used to refresh cache.
+	GceService AutoscalingGceClient
+}
+
+// MigInformation is a wrapper for Mig.
+type MigInformation struct {
+	Config   Mig
+	Basename string
+}
+
+// MachineTypeKey is used to identify MachineType.
+type MachineTypeKey struct {
+	Zone        string
+	MachineType string
+}
+
+// NewGceCache creates empty GceCache.
+func NewGceCache(gceService AutoscalingGceClient) GceCache {
+	return GceCache{
+		migs:          []*MigInformation{},
+		migCache:      map[GceRef]Mig{},
+		machinesCache: map[MachineTypeKey]*gce.MachineType{},
+		GceService:    gceService,
+	}
+}
+
+//  Methods locking on migsMutex.
+
+// RegisterMig returns true if the node group wasn't in cache before, or its config was updated.
+func (gc *GceCache) RegisterMig(mig Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	for i := range gc.migs {
+		if oldMig := gc.migs[i].Config; oldMig.GceRef() == mig.GceRef() {
+			if !reflect.DeepEqual(oldMig, mig) {
+				gc.migs[i].Config = mig
+				glog.V(4).Infof("Updated Mig %s", mig.GceRef().String())
+				return true
+			}
+			return false
+		}
+	}
+
+	glog.V(1).Infof("Registering %s", mig.GceRef().String())
+	gc.migs = append(gc.migs, &MigInformation{
+		Config: mig,
+	})
+	return true
+}
+
+// UnregisterMig returns true if the node group has been removed, and false if it was alredy missing from cache.
+func (gc *GceCache) UnregisterMig(toBeRemoved Mig) bool {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	newMigs := make([]*MigInformation, 0, len(gc.migs))
+	found := false
+	for _, mig := range gc.migs {
+		if mig.Config.GceRef() == toBeRemoved.GceRef() {
+			glog.V(1).Infof("Unregistered Mig %s", toBeRemoved.GceRef().String())
+			found = true
+		} else {
+			newMigs = append(newMigs, mig)
+		}
+	}
+	gc.migs = newMigs
+	return found
+}
+
+// GetMigs returns a copy of migs list.
+func (gc *GceCache) GetMigs() []*MigInformation {
+	gc.migsMutex.Lock()
+	defer gc.migsMutex.Unlock()
+
+	migs := make([]*MigInformation, 0, len(gc.migs))
+	for _, mig := range gc.migs {
+		migs = append(migs, &MigInformation{
+			Basename: mig.Basename,
+			Config:   mig.Config,
```

Comment:
It seems like the semantic here is changing rather significantly. Previously MigInformation had a Mig struct, but now it has Mig interface implemented by *gceMig. Are we implicitly moving from deep to shallow copy here?

Reply:
Right, I thought it was a field, because I was going commit-by-commit through the review and not looking at the whole thing (because of how large it is :P). I think it *should* be a deep copy, but that would require a more careful consideration. Let's leave it as is for this PR and think if we need a follow up making it a deepcopy (note - it's probably only relevant with autoprovisioning, so it may not be as battle-tested as we may think it is).